numpyro/util.py

```diff
@@ -104,3 +109,103 @@ def scan(f, a, bs):
         return tree_unflatten(tree_def, as_)
     else:
         return lax.scan(f, a, bs)
+
+
+def tscan(f, a, bs, fields=(0,)):
+    if _DISABLE_CONTROL_FLOW_PRIM:
+        # convert pytree to flat jaxtuple
+        a, a_tree = pytree_to_flatjaxtuple(a)
+        bs, b_tree = pytree_to_flatjaxtuple(bs)
+        length = tuple(bs)[0].shape[0]
+        state = [lax.full((length,) + a[i].shape, 0, lax._dtype(a[i])) for i in fields]
+
+        def body_fun(i, vals):
+            a, state = vals
+            # select i-th element from each b
+            b = [b[i] for b in bs]
+            ff, _ = pytree_fun_to_flatjaxtuple_fun(wrap_init(f), (a_tree, b_tree))
+            a_out = ff.call_wrapped(a, core.pack(b))
```

Comment:
Somehow we can't use this pattern 2 times; so ff is created for each iteration. In addition, there is no documentation in jax to show how to manipulate these stuffs. :(

Reply:
Maybe I'm missing something, but this seems quite a bit more complicated. Would something like this not work?

```python
def tscan(f, a, bs, fields=(0,)):
    if _DISABLE_CONTROL_FLOW_PRIM:
        bs, pack_fn = ravel_pytree(bs)
        as_ = []
        _, tree_def = tree_flatten(bs)
        for b in bs:
            as_.append([x if i in fields else None for i, x in enumerate(f(a, b))])
        return tree_unflatten(tree_def, as_)
    else:
        return _tscan(f, a, bs)
```

Note that since this is only for debugging, we don't have to care about efficiency. Maybe you had something else in mind though?

Reply:
This can work for a simple version of lax.scan, where a is a simple pytree, b is a simple pytree. But it seems does not work for a simple example as in the test (I remember I get some errors on the shape mismatch).